File: apps/api/src/modules/checkin/checkin.service.ts
import { Inject, Injectable } from "@nestjs/common";
import type { CheckinCreateInput, CheckinDto } from "@today/contracts";
import { IdentityService } from "../identity/identity.service";

/**
 * checkin — 每日记录写入/读取。
 * 持久化（Prisma）下一步接入；当前内存实现仅用于模块边界可运行。
 */
@Injectable()
export class CheckinService {
  private readonly store = new Map<string, CheckinDto>();

  constructor(
    @Inject(IdentityService) private readonly identity: IdentityService,
  ) {}

  private key(userId: string, date: string) {
    return `${userId}:${date}`;
  }

  todayDate() {
    return new Date().toISOString().slice(0, 10);
  }

  getToday(): CheckinDto | null {
    const userId = this.identity.getCurrentUserId();
    return this.store.get(this.key(userId, this.todayDate())) ?? null;
  }

  upsert(input: CheckinCreateInput): CheckinDto {
    const userId = this.identity.getCurrentUserId();
    const date = input.date ?? this.todayDate();
    const now = new Date().toISOString();
    const existing = this.store.get(this.key(userId, date));
    const dto: CheckinDto = {
      id: existing?.id ?? crypto.randomUUID(),
      userId,
      date,
      rawText: input.rawText,
      createdAt: existing?.createdAt ?? now,
      updatedAt: now,
    };
    this.store.set(this.key(userId, date), dto);
    return dto;
  }

  listRecent(limit = 30): CheckinDto[] {
    const userId = this.identity.getCurrentUserId();
    return [...this.store.values()]
      .filter((c) => c.userId === userId)
      .sort((a, b) => b.date.localeCompare(a.date))
      .slice(0, limit);
  }
}

File: apps/api/src/main/java/com/today/checkin/CheckinService.java
package com.today.checkin;

import com.today.common.EntityMapper;
import com.today.identity.IdentityService;
import com.today.persistence.CheckinEntity;
import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class CheckinService {

  private final CheckinMapper checkinMapper;
  private final IdentityService identity;

  public CheckinService(CheckinMapper checkinMapper, IdentityService identity) {
    this.checkinMapper = checkinMapper;
    this.identity = identity;
  }

  public String todayDate() {
    return EntityMapper.todayUtc().toString();
  }

  public CheckinDto getToday() {
    CheckinEntity entity =
        checkinMapper.findByUserIdAndDate(identity.getCurrentUserId(), EntityMapper.todayUtc());
    return entity == null ? null : EntityMapper.toDto(entity);
  }

  @Transactional
  public CheckinDto upsert(CheckinCreateInput input) {
    String userId = identity.getCurrentUserId();
    LocalDate date =
        input.date() != null ? EntityMapper.parseDate(input.date()) : EntityMapper.todayUtc();
    Instant now = EntityMapper.now();

    CheckinEntity existing = checkinMapper.findByUserIdAndDate(userId, date);
    if (existing == null) {
      CheckinEntity created = new CheckinEntity();
      created.setId(UUID.randomUUID().toString());
      created.setUserId(userId);
      created.setCheckinDate(date);
      created.setRawText(input.rawText().trim());
      created.setCreatedAt(now);
      created.setUpdatedAt(now);
      checkinMapper.insert(created);
      return EntityMapper.toDto(created);
    }

    existing.setRawText(input.rawText().trim());
    existing.setUpdatedAt(now);
    checkinMapper.update(existing);
    return EntityMapper.toDto(existing);
  }

  public List<CheckinDto> listRecent(int limit) {
    return checkinMapper.listRecentByUserId(identity.getCurrentUserId(), limit).stream()
        .map(EntityMapper::toDto)
        .toList();
  }

  /** 流水线内部按 id 读取（不校验当前用户，调用方已持有 job） */
  public CheckinDto findByIdForPipeline(String checkinId) {
    CheckinEntity entity = checkinMapper.findById(checkinId);
    return entity == null ? null : EntityMapper.toDto(entity);
  }
}
